engine/jobmaster/dm/dm_jobmaster.go

```diff
@@ -188,8 +188,12 @@ func (jm *JobMaster) onWorkerFinished(taskStatus runtime.TaskStatus, worker lib.
 
 // OnWorkerStatusUpdated implements JobMasterImpl.OnWorkerStatusUpdated
 func (jm *JobMaster) OnWorkerStatusUpdated(worker lib.WorkerHandle, newStatus *libModel.WorkerStatus) error {
-	// No need to do anything here, because we update it in OnWorkerOnline
-	return nil
+	// we alreay update finished status in OnWorkerOffline
+	if newStatus.Code == libModel.WorkerStatusFinished || len(newStatus.ExtBytes) == 0 {
+		return nil
+	}
+	log.L().Debug("on worker status updated", zap.String("extra bytes", string(newStatus.ExtBytes)), zap.String("id", jm.workerID), zap.String("worker_id", worker.ID()))
+	return jm.OnWorkerOnline(worker)
```

Comment:
why we treat it (such as task paused by error) as WorkerOnline? maybe we can extract updateStatus, then let OnWorkerStatusUpdated and OnWorkerOnline to call it.

Reply:
for jobmaster, it only need to know whether the worker is online or offline. TaskManager operate task when worker online, WorkerManager scheduler worker when worker offline. For a task status updated, the worker still online and we can operate for it, so I think we should rely on OnWorkerOnline.

Reply:
Directly let OnWorkerStatusUpdated call OnWorkerOnline is strange. The two functions are callbacks of framework, and have very clear meaning. I still prefer we extract our private function and let them two use it.

Reply:
OK, we can left a comment to explain it

Reply:
I am now thinking a debug scenario. When I see the logic is inside OnWorkerOnline, I may skip checking the caller because the meaning of OnWorkerOnline is very clear, then I jump to the logic of dmWorker online. But the bug may lie in dmWorker update stage.

The point is we want to share some code in the two callback, but the callback has very clear definition that indicates its usage, so in my mind even though the code is same the two functions are still separate functions, they should not be related to each other.

Reply:
done